### ISSUE TITLE
[Block editor] Shortcodes and frontend functions can lead to fatal

### ISSUE THREAD

**Issue by barryhughes** [woothemes/woocommerce#38169]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Simple shortcodes that take advantage of various WooCommerce helper functions such as `wc_print_notices()` (from within the shortcode's callback function) fail in the context of the block editor. Minimal example:

```php
add_shortcode( 'my_test', function () {
	ob_start();
	wc_print_notices();
	echo '<div>MY CUSTOM SHORTCODE</div>';
	return ob_get_clean();
} );
```

If a new page is created, and it contains a 'problematic' shortcode like the one defined above, the page cannot be saved as we will see a **Publishing Failed** error (see following screenshot). If an *existing* post is edited, and a problematic shortcode is added, we will essentially hit the same problem except the error will read **Updating Failed**.

<div align="center"><img width="700" alt="publishing-failed" src="https://user-images.githubusercontent.com/3594411/236886155-39070556-597d-402d-938a-2bc0cc176ed2.png"></div>

### Expected behavior

No such error should occur, and the page should be successfully be published or updated.

### Actual behavior

The described error will occur, and the page or post cannot be saved or updated until the shortcode is removed. Again, this is only an issue in the block editor.

### Steps to reproduce

1. Add the sample code provided above.
2. Create a post or page (this must be done using the block editor).
3. Add the test shortcode `[my_test]` via a shortcode block, and then try to save.
4. Observe the error.
5. Rinse and repeat in the 'classic' editor, there is no problem.

### Notes

The problem here is two-fold:

- Various functions such as `wc_print_notices()` are [only loaded in the context of frontend requests](https://github.com/woocommerce/woocommerce/blob/7.6.1/plugins/woocommerce/includes/class-woocommerce.php#L553-L555).
- However, when a post containing a shortcode block is saved, the handler/callback is invoked outside of that context, resulting in undefined function errors.

I've logged this as a bug, because it feels like changing how and when WooCommerce Core includes these functions is the least disruptive way to solve this going forward—but an alternative would be to take the stance that existing any existing shortcodes that are subject to this problem should be updated and take their own precautions.

### WordPress Environment

—

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by barryhughes** [comment]
Essentially the same problem is also being tracked over here:

https://github.com/WordPress/gutenberg/issues/18988

**Comment by rrennick** [comment]
@barryhughes Shortcodes should only return strings and not generate any output:

```
add_shortcode( 'my_test', function () {
	ob_start();
	wc_print_notices();
	return ob_get_clean() . '<div>MY CUSTOM SHORTCODE</div>';
} );
```

**Comment by barryhughes** [comment]
Hey @rrennick!

> Shortcodes should only return strings and not generate any output:

Correct, thank you—I updated the snippet in the description (but, the core problem remains the same, even if we use output buffering to wrap the calls to the potentially undefined functions).

**Comment by kaushikasomaiya** [comment]
6607484-zen

replicated by using WC Store notices default widget in Appearance > Widgets area.

**Comment by Brianmitchtay** [comment]
I may be seeing another case of this in 7124022-zen

```
2023-10-05T00:11:04+00:00 CRITICAL Uncaught Error: Call to undefined function wc_print_notices() in /wordpress/plugins/woocommerce-accelerated-frequency/8.2.0.30/includes/wc-template-functions.php:3878
Stack trace:
#0 /wordpress/plugins/woocommerce-accelerated-frequency/8.2.0.30/packages/woocommerce-blocks/src/BlockTypes/StoreNotices.php(30): woocommerce_output_all_notices()
#1 /wordpress/plugins/woocommerce-accelerated-frequency/8.2.0.30/packages/woocommerce-blocks/src/BlockTypes/AbstractBlock.php(88): Automattic\WooCommerce\Blocks\BlockTypes\StoreNotices->render(Array, '', Object(WP_Block))
#2 /wordpress/core/6.3.1/wp-includes/class-wp-block.php(258): Automattic\WooCommerce\Blocks\BlockTypes\AbstractBlock->render_callback(Array, '', Object(WP_Block))
#3 /wordpress/core/6.3.1/wp-includes/blocks.php(1133): WP_Block->render()
#4 /wordpress/core/6.3.1/wp-includes/blocks.php(1171): render_block(Array)
#5 /wordpress/core/6.3.1/wp-includes/class-wp-hook.php(310): do_blocks('<!-- wp:group {...')
#6 /wordpress/core/6.3.1/wp-includes/plugin.php(205): WP_Hook->apply_filters('<!-- wp:group {...', Array)
#7 /wordpress/core/6.3.1/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php(1864): apply_filters('the_content', '<!-- wp:group {...')
#8 /wordpress/core/6.3.1/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php(385): WP_REST_Posts_Controller->prepare_item_for_response(Object(WP_Post), Object(WP_REST_Request))
#9 /wordpress/core/6.3.1/wp-includes/rest-api/class-wp-rest-server.php(1188): WP_REST_Posts_Controller->get_items(Object(WP_REST_Request))
#10 /wordpress/core/6.3.1/wp-includes/rest-api/class-wp-rest-server.php(1035): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/wp/v2/pages', Array, NULL)
#11 /wordpress/core/6.3.1/wp-includes/rest-api/class-wp-rest-server.php(447): WP_REST_Server->dispatch(Object(WP_REST_Request))
#12 /wordpress/core/6.3.1/wp-includes/rest-api.php(418): WP_REST_Server->serve_request('/wp/v2/pages')
#13 /wordpress/core/6.3.1/wp-includes/class-wp-hook.php(310): rest_api_loaded(Object(WP))
#14 /wordpress/core/6.3.1/wp-includes/class-wp-hook.php(334): WP_Hook->apply_filters(NULL, Array)
#15 /wordpress/core/6.3.1/wp-includes/plugin.php(565): WP_Hook->do_action(Array)
#16 /wordpress/core/6.3.1/wp-includes/class-wp.php(398): do_action_ref_array('parse_request', Array)
#17 /wordpress/core/6.3.1/wp-includes/class-wp.php(779): WP->parse_request('')
#18 /wordpress/core/6.3.1/wp-includes/functions.php(1335): WP->main('')
#19 /wordpress/core/6.3.1/wp-blog-header.php(16): wp()
#20 /wordpress/core/6.3.1/index.php(17): require('/wordpress/core...')
#21 {main}
  thrown in /wordpress/plugins/woocommerce-accelerated-frequency/8.2.0.30/includes/wc-template-functions.php on line 3878
```

When loading the affected page (homepage) in the editor. It whitescreens the entire page, since that page is loaded in the initial load of the pages area of the template editor, it also seems to prevent any pages from showing in the sidebar list of that view. The above error is logged in that case 

I've not been able to replicate the issue in testing thus far because I'm not able to load the page in the editor to see what shortcode is causing the issue. The page loads fine on the front-end though.

It's entirely possible the issue I'm seeing isn't exactly what's been described here, but it's too close for me to not note it.

**Comment by barryhughes** [comment]
Hi @Brianmitchtay,

> I've not been able to replicate the issue in testing thus far because I'm not able to load the page in the editor to see what shortcode is causing the issue.

- This plus the reference to `wc_print_notices()` in the backtrace makes me think it is indeed the same issue. 
- https://github.com/woocommerce/woocommerce/pull/40648 (fix) has been merged and should become available in 8.4.0 (scheduled for December 12).
- If a temporary mitigation is needed before then, it should probably work if you add the following code to a location such as `wp-content/mu-plugins/temp-fix-wc-shortcodes.php`:

```php
<?php
/**
 * Remove after 8.4.0.
 *
 * @see https://github.com/woocommerce/woocommerce/issues/38169
 */
add_action( 'before_woocommerce_init', function () {
	wc()->frontend_includes();
} );
```
